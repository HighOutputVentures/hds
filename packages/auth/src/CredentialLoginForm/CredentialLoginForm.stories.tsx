import { ThemeProvider } from "@highoutput/hds";
import { ComponentStory, Meta } from "@storybook/react";
import * as React from "react";
import CredentialLoginForm from "./CredentialLoginForm";

const Story: Meta<typeof CredentialLoginForm> = {
  title: "Templates/Login/CredentialLoginForm",
  component: CredentialLoginForm,
};

export default Story;

const Template: ComponentStory<typeof CredentialLoginForm> = (args) => (
  <ThemeProvider>
    <CredentialLoginForm maxW="400px" {...args} />
  </ThemeProvider>
);

export const Default = Template.bind({});

Default.args = {
  ...Default.args,
  variant: "email-password",
  onSubmit(data) {
    alert(JSON.stringify(data));
  },
};
