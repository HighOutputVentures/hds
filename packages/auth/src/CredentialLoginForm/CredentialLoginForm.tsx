import * as React from "react";
import { FC } from "react";
import { omit } from "../utils";
import {
  EmailPasswordLoginForm,
  EmailPasswordLoginFormProps,
} from "./EmailPasswordLoginForm";
import {
  UsernamePasswordLoginForm,
  UsernamePasswordLoginFormProps,
} from "./UsernamePasswordLoginForm";

export type CredentialLoginFormProps =
  | ({ variant: "name-password" } & UsernamePasswordLoginFormProps)
  | ({ variant: "email-password" } & EmailPasswordLoginFormProps);

export const CredentialLoginForm: FC<CredentialLoginFormProps> = (props) => {
  switch (props.variant) {
    case "email-password":
      return <EmailPasswordLoginForm {...omit(props, "variant")} />;
    case "name-password":
      return <UsernamePasswordLoginForm {...omit(props, "variant")} />;
    default:
      return null;
  }
};

export default CredentialLoginForm;
