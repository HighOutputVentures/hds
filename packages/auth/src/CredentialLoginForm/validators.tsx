import * as yup from "yup";

export const UsernamePasswordLoginSchema = yup
  .object({
    username: yup.string().required("Username is required"),
    password: yup.string().required("Password is required"),
  })
  .required();

export type TUsernamePasswordLoginSchema = yup.InferType<
  typeof UsernamePasswordLoginSchema
>;

export const EmailPasswordLoginSchema = yup
  .object({
    emailAddress: yup
      .string()
      .email("Invalid email address")
      .required("Email address is required"),
    password: yup.string().required("Password is required"),
  })
  .required();

export type TEmailPasswordLoginSchema = yup.InferType<typeof EmailPasswordLoginSchema>;
