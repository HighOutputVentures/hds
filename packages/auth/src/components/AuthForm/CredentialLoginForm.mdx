import { Canvas, Meta, Story, ArgsTable } from '@storybook/addon-docs';

import StorybookHeader from '../../../../../components/StorybookHeader';

<Meta title="Components/Auth Form/Credential Login" />

<StorybookHeader
  title="HDS Authentication"
  command="npm install @highoutput/hds-auth"
/>

<Canvas>
  <Story id="components-auth-form-credential-login--default"></Story>
</Canvas>

<ArgsTable />
