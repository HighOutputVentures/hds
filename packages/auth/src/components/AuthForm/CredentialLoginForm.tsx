import { ViewIcon, ViewOffIcon } from '@chakra-ui/icons';
import { Box, Button, Center, Stack, Text } from '@chakra-ui/react';
// @ts-ignore
import { InputField } from '@highoutput/hds-forms';
import { yupResolver } from '@hookform/resolvers/yup';
import React, { FC, ReactNode } from 'react';
import { useForm } from 'react-hook-form';

import {
  CredentialFormInputEmailProps,
  CredentialFormInputNameProps,
  withCredentialFormSchemaEmail,
  withCredentialFormSchemaEmailValues,
  withCredentialFormSchemaName,
  withCredentialFormSchemaNameValues,
} from './validation';

export type CredentialLoginFormDefaultProps = {
  loginTitle?: ReactNode;
  signUpTitle?: ReactNode;
};
export interface CredentialLoginFormNameProps
  extends CredentialLoginFormDefaultProps {
  variant?: 'name-password';
  nameLabel: string;
  onSubmit?(values: CredentialFormInputNameProps): void;
}
export interface CredentialLoginFormEmailProps
  extends CredentialLoginFormDefaultProps {
  variant?: 'email-password';
  nameLabel?: never;
  onSubmit?(values: CredentialFormInputEmailProps): void;
}

export type CredentialLoginFormProps =
  | CredentialLoginFormNameProps
  | CredentialLoginFormEmailProps;

const CredentialLoginForm: FC<CredentialLoginFormProps> = (props) => {
  const {
    signUpTitle,
    loginTitle,
    variant,
    onSubmit,
    nameLabel = 'Username',
  } = props;

  const [showPassword, setShowPassword] = React.useState(false);
  const [isSignUp, setIsSignUp] = React.useState(false);
  const { register, handleSubmit, formState } = useForm<
    withCredentialFormSchemaEmailValues & withCredentialFormSchemaNameValues
  >({
    resolver: yupResolver(
      variant === 'name-password'
        ? withCredentialFormSchemaName
        : withCredentialFormSchemaEmail
    ),
    shouldUnregister: true,
  });

  const onSubmitForm = async (
    values: withCredentialFormSchemaNameValues &
      withCredentialFormSchemaEmailValues
  ) => {
    if (onSubmit) {
      onSubmit(values);
    }
  };

  return (
    <Box as={'form'} maxW={512} onSubmit={handleSubmit(onSubmitForm)}>
      <Center m={0} p={0}>
        {isSignUp && signUpTitle ? (
          signUpTitle
        ) : !isSignUp && loginTitle ? (
          loginTitle
        ) : !isSignUp && !loginTitle ? (
          <Text size="header-3" my={8}>
            Login
          </Text>
        ) : (
          isSignUp &&
          !signUpTitle && (
            <Text size="header-3" my={8}>
              Sign up
            </Text>
          )
        )}
      </Center>

      <Stack spacing={'1rem'}>
        {variant === 'name-password' ? (
          <InputField
            {...register('name')}
            id={'name'}
            label={nameLabel?.charAt(0).toUpperCase() + nameLabel?.slice(1)}
            placeholder={`Input your ${nameLabel?.toLowerCase()}`}
            errorMsg={formState.errors.name?.message}
            disabled={formState.isSubmitting}
            // partProps={{
            //   input: {
            //     'aria-label': 'name-input',
            //     role: 'input',
            //   },
            // }}
          />
        ) : (
          <InputField
            {...register('email')}
            id={'email'}
            label="Email"
            placeholder="Input your email"
            errorMsg={formState.errors.email?.message}
            disabled={formState.isSubmitting}
            // partProps={{
            //   input: { 'aria-label': 'email-input', role: 'input' },
            // }}
          />
        )}

        <InputField
          {...register('password')}
          id="password"
          label="Password"
          type={showPassword ? 'text' : 'password'}
          placeholder="Input your password"
          errorMsg={formState.errors.password?.message}
          disabled={formState.isSubmitting}
          rightIcon={
            <Button
              data-testid={'show-hide-btn'}
              background={'none'}
              _hover={{ background: 'none' }}
              _active={{ background: 'none' }}
              aria-label={showPassword ? 'show-password' : 'hide-password'}
              onClick={() => setShowPassword((prev) => !prev)}
            >
              {showPassword ? <ViewIcon /> : <ViewOffIcon />}
            </Button>
          }
          // partProps={{
          //   input: {
          //     'aria-label': 'password-input',
          //     role: 'input',
          //   },
          // }}
        />
      </Stack>
      <Button
        w="full"
        variant="primary"
        isLoading={formState.isSubmitting}
        type="submit"
        my={8}
      >
        {isSignUp ? 'Sign Up' : 'Login'}
      </Button>
      <Center>
        <Text>
          {isSignUp ? 'Already have an account?' : 'No account yet?'}{' '}
          <Text
            as={'a'}
            data-testid={'switch-form-link'}
            _hover={{ textDecoration: 'underline', cursor: 'pointer' }}
            aria-label={isSignUp ? 'login-link-label' : 'signup-link-label'}
            onClick={() => setIsSignUp((prev) => !prev)}
            fontWeight={'bold'}
            role={'link'}
          >
            {isSignUp ? 'Login' : 'Sign up'}
          </Text>
        </Text>
      </Center>
    </Box>
  );
};

export default CredentialLoginForm;
