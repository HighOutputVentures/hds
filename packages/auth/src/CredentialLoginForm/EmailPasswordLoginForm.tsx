import { chakra, Stack, SystemStyleObject } from '@chakra-ui/react';
import { Button, PasswordField, TextField } from '@highoutput/hds-forms';
import { yupResolver } from '@hookform/resolvers/yup';
import * as React from 'react';
import { useForm } from 'react-hook-form';
import { EmailPasswordLoginSchema, TEmailPasswordLoginSchema } from './validators';

type FieldConfig = {
  label?: string;
};

type ButtonConfig = {
  label?: string;
};

export type EmailPasswordLoginFormProps = Omit<SystemStyleObject, 'onSubmit'> & {
  emailAdd?: FieldConfig;
  password?: FieldConfig;
  submitBtn?: ButtonConfig;
  renderFooter?: React.ReactNode;
  renderHeader?: React.ReactNode;
  onSubmit?(values: TEmailPasswordLoginSchema): void;
  __emailAddTestId?: string;
  __passwordTestId?: string;
  __submitBtnTestId?: string;
};

export const EmailPasswordLoginForm: React.FC<EmailPasswordLoginFormProps> = ({
  emailAdd,
  password,
  submitBtn,
  renderHeader,
  renderFooter,
  onSubmit,

  __emailAddTestId = 'hds.credential-login.input.email',
  __passwordTestId = 'hds.credential-login.input.password',
  __submitBtnTestId = 'hds.credential-login.submit-button',

  ...others
}) => {
  const { register, formState, handleSubmit } = useForm<TEmailPasswordLoginSchema>({
    shouldFocusError: true,
    shouldUnregister: true,
    resolver: yupResolver(EmailPasswordLoginSchema),
    defaultValues: {
      password: '',
      emailAddress: '',
    },
  });

  return (
    <chakra.div
      sx={{
        maxW: '400px',
        ...others,
      }}
      data-testid="hds.credential-login"
    >
      {renderHeader}

      <chakra.form onSubmit={handleSubmit(onSubmit ?? function noop() {})}>
        <Stack spacing={4}>
          <TextField
            id="email"
            label={emailAdd?.label}
            placeholder="Input your email"
            error={formState.errors.emailAddress?.message}
            isDisabled={formState.isSubmitting}
            __fieldTestId={__emailAddTestId}
            {...register('emailAddress')}
          />

          <PasswordField
            id="password"
            label={password?.label}
            placeholder="Input your password"
            error={formState.errors.password?.message}
            isDisabled={formState.isSubmitting}
            __fieldTestId={__passwordTestId}
            {...register('password')}
          />
        </Stack>

        <Button
          type="submit"
          width="full"
          marginTop={8}
          isLoading={formState.isSubmitting}
          __testId={__submitBtnTestId}
        >
          {submitBtn?.label ?? 'Login'}
        </Button>
      </chakra.form>

      {renderFooter}
    </chakra.div>
  );
};
