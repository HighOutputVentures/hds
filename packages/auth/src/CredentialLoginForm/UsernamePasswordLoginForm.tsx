import { chakra, Stack, SystemStyleObject } from '@chakra-ui/react';
import { Button, PasswordField, TextField } from '@highoutput/hds-forms';
import { yupResolver } from '@hookform/resolvers/yup';
import * as React from 'react';
import { useForm } from 'react-hook-form';
import { TUsernamePasswordLoginSchema, UsernamePasswordLoginSchema } from './validators';

type FieldConfig = {
  label?: string;
};

type ButtonConfig = {
  label?: string;
};

export type UsernamePasswordLoginFormProps = Omit<SystemStyleObject, 'onSubmit'> & {
  username?: FieldConfig;
  password?: FieldConfig;
  submitBtn?: ButtonConfig;
  renderFooter?: React.ReactNode;
  renderHeader?: React.ReactNode;
  onSubmit?(values: TUsernamePasswordLoginSchema): void;
  __usernameTestId?: string;
  __passwordTestId?: string;
  __submitBtnTestId?: string;
};

export const UsernamePasswordLoginForm: React.FC<UsernamePasswordLoginFormProps> = ({
  username,
  password,
  submitBtn,
  renderHeader,
  renderFooter,
  onSubmit,

  __usernameTestId = 'hds.credential-login.input.username',
  __passwordTestId = 'hds.credential-login.input.password',
  __submitBtnTestId = 'hds.credential-login.submit-button',

  ...others
}) => {
  const { register, formState, handleSubmit } = useForm<TUsernamePasswordLoginSchema>({
    shouldFocusError: true,
    shouldUnregister: true,
    resolver: yupResolver(UsernamePasswordLoginSchema),
    defaultValues: {
      username: '',
      password: '',
    },
  });

  return (
    <chakra.div
      sx={{
        maxW: '400px',
        ...others,
      }}
      data-testid="hds.credential-login"
    >
      {renderHeader}

      <chakra.form onSubmit={handleSubmit(onSubmit ?? function noop() {})}>
        <Stack spacing={4}>
          <TextField
            label={username?.label}
            placeholder="Input your username"
            error={formState.errors.username?.message}
            isDisabled={formState.isSubmitting}
            __fieldTestId={__usernameTestId}
            {...register('username')}
          />

          <PasswordField
            id="password"
            label={password?.label}
            placeholder="Input your password"
            error={formState.errors.password?.message}
            isDisabled={formState.isSubmitting}
            __fieldTestId={__passwordTestId}
            {...register('password')}
          />
        </Stack>

        <Button
          type="submit"
          width="full"
          marginTop={8}
          isLoading={formState.isSubmitting}
          __testId={__submitBtnTestId}
        >
          {submitBtn?.label ?? 'Login'}
        </Button>
      </chakra.form>

      {renderFooter}
    </chakra.div>
  );
};
