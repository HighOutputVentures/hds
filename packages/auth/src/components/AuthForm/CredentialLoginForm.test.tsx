import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import CredentialLoginForm from './CredentialLoginForm';

describe('Login Form Email-based Component', () => {
  beforeEach(() => {
    render(<CredentialLoginForm variant="email-password" />);
  });
  it('should display error message on email', async () => {
    // const emailInput = screen.getByRole('input', { name: 'email-input' });
    const emailInput = screen.getByTestId('hds.credential.input.email');
    const submitBtn = screen.getByRole('button', { name: 'Login' });
    fireEvent.change(emailInput, { target: { value: 'hello@g' } });
    fireEvent.click(submitBtn);
    const errorText = await screen.findByText(
      /please enter a valid email address/i
    );
    expect(errorText).toBeInTheDocument();
  });

  it('should display error message on password', async () => {
    // const passwordInput = screen.getByRole('input', { name: 'password-input' });
    const passwordInput = screen.getByPlaceholderText(/Input your password/i);
    const submitBtn = screen.getByRole('button', { name: 'Login' });
    fireEvent.change(passwordInput, { target: { value: null } });
    fireEvent.click(submitBtn);
    const errorText = await screen.findByText(/password is required/i);
    expect(errorText).toBeInTheDocument();
  });
  it('should toggle login to signup form', async () => {
    const switchLink = screen.getByTestId('hds.switch.form.link');
    fireEvent.click(switchLink);
    const textLoginLabel = screen.getByRole('link', {
      name: 'hds.login.link.label',
    });
    expect(textLoginLabel).toBeInTheDocument();
    fireEvent.click(switchLink);
    const textSignUpLabel = await screen.findByRole('link', {
      name: 'hds.signup.link.label',
    });
    expect(textSignUpLabel).toBeInTheDocument();
  });
});

describe('Login form Username-based component', () => {
  beforeEach(() => {
    render(
      <CredentialLoginForm variant="name-password" nameLabel="Username" />
    );
  });
  it('should display error message on name', async () => {
    // const nameInput = screen.getByRole('input', { name: 'name-input' });
    const nameInput = screen.getByTestId('hds.credential.input.username');
    const submitBtn = screen.getByRole('button', { name: 'Login' });
    fireEvent.change(nameInput, { target: { value: null } });
    fireEvent.click(submitBtn);
    const errorText = await screen.findByText(/name is required/i);
    expect(errorText).toBeInTheDocument();
  });
});

describe('Snapshot', () => {
  it('(CredentialLoginForm) Should match snapshot', () => {
    const component = renderer.create(
      <CredentialLoginForm variant="name-password" />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    component.unmount();
  });
});
