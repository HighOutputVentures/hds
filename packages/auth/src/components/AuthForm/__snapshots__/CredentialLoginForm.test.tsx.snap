// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot (CredentialLoginForm) Should match snapshot 1`] = `
<form
  className="css-1er7eb5"
  onSubmit={[Function]}
>
  <div
    className="css-1gwedk2"
  >
    <div
      className="chakra-stack css-1it59rx"
    >
      <p
        className="chakra-text css-0"
      >
        Login
      </p>
    </div>
  </div>
  <div
    className="chakra-stack css-sekf5y"
  >
    <div
      aria-label="Form Group"
      className="chakra-form-control css-0"
      data-testid=":r1:-form-container-form-control"
      role="group"
    >
      <div
        className="chakra-input__group css-4302v8"
        data-testid=":r0:-input-field-group"
      >
        <input
          className="chakra-input css-0"
          data-testid=":r0:-input-field-input"
          disabled={false}
          id="name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Input your username"
          readOnly={false}
          required={false}
          role="input"
          type="text"
        />
      </div>
    </div>
    <div
      aria-label="Form Group"
      className="chakra-form-control css-0"
      data-testid=":r4:-form-container-form-control"
      role="group"
    >
      <div
        className="chakra-input__group css-4302v8"
        data-testid=":r3:-input-field-group"
      >
        <input
          className="chakra-input css-0"
          data-testid=":r3:-input-field-input"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Input your password"
          readOnly={false}
          required={false}
          role="input"
          type="password"
        />
        <div
          className="chakra-input__right-element css-dp9r5f"
          data-testid=":r3:-input-field-right-element"
          size="md"
        >
          <button
            aria-label="hide-password"
            className="chakra-button css-hx0lo2"
            data-testid="show-hide-btn"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="chakra-icon css-17kjh4s"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M23.2,10.549a20.954,20.954,0,0,0-4.3-3.6l4-3.995a1,1,0,1,0-1.414-1.414l-.018.018a.737.737,0,0,1-.173.291l-19.5,19.5c-.008.007-.018.009-.026.017a1,1,0,0,0,1.631,1.088l4.146-4.146a11.26,11.26,0,0,0,4.31.939h.3c4.256,0,8.489-2.984,11.051-5.8A2.171,2.171,0,0,0,23.2,10.549ZM16.313,13.27a4.581,4.581,0,0,1-3,3.028,4.3,4.3,0,0,1-3.1-.19.253.253,0,0,1-.068-.407l5.56-5.559a.252.252,0,0,1,.407.067A4.3,4.3,0,0,1,16.313,13.27Z"
                />
                <path
                  d="M7.615,13.4a.244.244,0,0,0,.061-.24A4.315,4.315,0,0,1,7.5,12,4.5,4.5,0,0,1,12,7.5a4.276,4.276,0,0,1,1.16.173.244.244,0,0,0,.24-.062l1.941-1.942a.254.254,0,0,0-.1-.421A10.413,10.413,0,0,0,12,4.75C7.7,4.692,3.4,7.7.813,10.549a2.15,2.15,0,0,0-.007,2.9,21.209,21.209,0,0,0,3.438,3.03.256.256,0,0,0,.326-.029Z"
                />
              </g>
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <button
    className="chakra-button css-hfm3ds"
    disabled={false}
    type="submit"
  >
    Login
  </button>
  <div
    className="css-gmuwbf"
  >
    <p
      className="chakra-text css-0"
    >
      No account yet?
       
      <a
        aria-label="signup-link-label"
        className="chakra-text css-1xy2tp3"
        data-testid="switch-form-link"
        onClick={[Function]}
        role="link"
      >
        Sign up
      </a>
    </p>
  </div>
</form>
`;
